### DISCUSSION TITLE
Minimal version for different flash layout

### DISCUSSION THREAD

**Discussion by protectivedad** [arendst/Tasmota#19823] [Support request] [answered]
I am using Tasmota (12.5) on a Globe 34324 light bulb. It was working well and I was changing my mqtt around to use TLS. At that time I noticed the bulbs have 2M flash, so I changed to `board = esp8266_2M1M` in my Platformio override file. I used the minimal Tasmota (12.5) downloaded from ota.tasmota.com to free space and OTA'd the new firmware (12.5). Worked great.

Now I want to update to 13.2. The problem is there is only 300 KB free program space. So I went and used minimal Tasmota (13.2) downloaded from ota.tasmota.com to free up space, At this point the bulb stopped working. I can not get access to it. Is this because of the changed flash layout? I've written off this bulb, but I don't want to go and ruin the other ones.

Should the tasmota minimal firmware (donwload from ota) work on an esp8266 board using the non traditional esp8266_2M1M, exp8266_2M256 and exp8266_4M? If I create a minimal firmware with a matching board definition will that work? At this point I am looking for an upgrade path for my existing esp8266_2M1M bulbs.

To help stop this problem from happening again I would like to go from Tasmota 12.5.0 (esp8266_2M1M) -> Tasmota 13.2.0 (esp8266_2M256).  That way I always have space to upgrade a lite/regular version.

**Comment by sfromis** [comment]
If you choose to use a binary build for a 2MB flash layout (it is not merely about size), you cannot then switch to a binary not built for 2MB. I'd recommend to stay away from doing that, unless you have very specific reasons for *needing* it, otherwise it is just more pain that it is worth, as it provides no benefit for normal usage. Easiest way to get out of your current mess is wired flashing of regular builds.

And 300KB of free space should be just fine for upgrading via a suitable binary build. I'd indeed expect you to need a minimal specifically built for 2MB flash layout.

**Comment by Jason2866** [comment]
If i where you, i would not try to update the other bulbs. The partition layouts needs to fit together from every firmware! So to update safe you have to know which version is currently installed and you have to build the needed minimal version with the same setting. If the device is back online with the minimal version you have flashed, the target new firmware has to be built with the same partition layout as used for the running minimal.
Every mixing can end in a brick.

**Comment by Jason2866** [comment]
For testing a way that may work, you can use a esp8266 devboard. There are no experience from us what's working or not.

**Reply by protectivedad** [reply]
I took apart the bulb I broke and have that to play around with. The bulb uses a TWE3L module. I will use that to see if I can safely upgrade. I wanted all the bulbs consistent so when I upgrade it simplifies things. Right now I have a mixture of 1M, 2M1M and 2M256. Which is not ideal.

**Comment by protectivedad** [comment, marked as answer]
One of my problems was I wasn't building the minimal versions. I was just using the downloaded minimal. Which, as was pointed out was a bad idea. I made a consistent set of 1M1M and 1M256 minimal/lite versions for 13.20.  With these my upgrade path was:
`12.5.0 1M1M -> minimal 13.2.0 1M1M -> lite 13.2.0 1M1M -> lite 13.2.0 1M256 -> custom 13.2.0 1M256`

Thanks for the information.
